This commit adds the extention types, datetime and duration to Cedar Go

apg opened the pull request:
See _pending_ [RFC 80](https://github.com/strongdm/cedar-rfcs/blob/datetime-rfc/text/0080-datetime-extension.md)

This PR adds operator overloading, by introducing a set of "Virtual"
comparison Evaler. Types that can overload operators must implement
the `Lesser` interface. Long, Datetime, and Duration are implemented in this
PR. (Note: It is expected that there will be a Decimal operator
overload RFC at some later date, but that is not hooked into this)

The now obsolete Evalers for Long comparisons have not been
removed. A possible improvement to this PR would be to utilize
them when it is known that Longs are on the right and left side.

The datetime itself is backed by an int64, and can be lifted from
a Go `time.Time`, or from integer and floats "unsafely." Duration, similarly, is backed by an int64, and can be lifted from a Go `time.Duration`, or from integer and floats "unsafely."

Commit:
This commit adds an extension type, datetime, to Cedar Go.

This PR does not add the duration type, nor the methods `toTime()`,
`offset()`, or `durationSince()` that would require it.

This PR adds operator overloading, by introducing a set of "Virtual"
comparison Evaler. Types that can overload operators must implement
the `Lesser` interface. Long and Datetime are implemented in this
PR. (Note: It is expected that there will be a Decimal operator
overload RFC at some later date, but that is not hooked into this)

The now obsolete Evalers for Long comparisons have not been
removed. A possible improvement to this PR would be to utilize
them when it is known that Longs are on the right and left side.

The datetime itself is backed by an int64, and can be lifted from
a Go `time.Time`, integer or float "unsafely."

Signed-off-by: Andrew Gwozdziewycz <andrew.gwozdziewycz@strongdm.com>

philhassey commented on internal/eval/evalers.go:
```diff
@@ -1073,3 +1120,123 @@ func (n *ipIsInRangeEval) Eval(ctx *Context) (types.Value, error) {
+		return zeroValue(), err
+	}
+	return types.Boolean(!lhs.Less(rhs)), nil
+}
```
I think it might need to be an error if the types are not the same.  Maybe this is part of the Less signature `Less(rhs) (bool, error)`

apg replied:
I agree that it should be an error. Happy to field suggestions for how to make that happen. I don't think a string comparison is it. Maybe the `Less` functions need to be `(bool, error)` since they actually can introspect values.

philhassey replied:
And I just realized, hold this thought for a bit.  There's a line of work that needs to happen on our end.  (We might be adding comparison methods to all value types that work as-per go `slices.SortFunc` requires.

philhassey replied:
There might even have to be two different comparisons, one for extension purposes, one for other purposes.

apg replied:
I'll leave for now.

philhassey replied:
One other (annoying) thought is that since the comparison error message depends on which is the LHS vs RHS we will either want two different methods (LessThan/MoreThan) or a single `Compare(rhs) int` so that we can be sure the type error complains about the RHS.

e.g. we want this 
```
5 < Duration("5s") // type error: expected long, got duration
5 > Duration("5s") // type error: expected long, got duration
```

not this
```
5 < Duration("5s") // type error: expected long, got duration
5 > Duration("5s") // type error: expected duration, got long
```

philhassey commented on types/datetime.go:
```diff
@@ -0,0 +1,292 @@
+	return Datetime{Value: int64(v)}
+}
+
+func FromStdTime(t time.Time) Datetime {
```
Maybe `DatetimeFromStandardTime` ... idk.

apg replied:
For Duration, I did `FromStdDuration` -- paired together, they make sense. We're lifting the "standard type" into Cedar Values. Could do the same for net.IP -- `FromStdIP`. 

I'm also flexible so don't really care one way or the other. :)

philhassey commented on types/datetime.go:
```diff
@@ -0,0 +1,292 @@
+	Value int64
+}
+
+func UnsafeDatetime[T int | int64 | float64](v T) Datetime {
```
Not sure what the utility of this one is.

apg replied:
It's been very useful in tests. I guess I don't truly see a need for it outside of that, so I can drop it.

philhassey commented on types/virtual.go:
```diff
@@ -0,0 +1,7 @@
+package types
+
+type Lesser interface {
```
For bikeshedding `Comparable`

apg replied:
`Comparable` goes against the Go conventions of being an "-er" and also breaks the convention of utilizing the name of the method if it makes sense. `Reader`, isn't `Funneler`, or `Gatherer` because the one method is called `Read`. :)

However, this is related to the above discussion of being able to support `slices.SortFunc` so, I'll defer to the outcome of that.

philhassey replied:
Yes and no, there's the `comparable` interface in Go now.  Though I think your suggestion is more correct.

https://go.dev/ref/spec#Type_constraints

All that said, it probably doesn't matter much, as I think we can have the `Lesser` interface be internal or private so it's not something the general public has to know about.

apg replied:
`comparable` is only for builtin types that have _all of the comparison_ operators defined for them. It's not a 1 method interface. :)

patjakdev replied:
As we discussed in DM, since this interface embeds `Value`, I think calling it `ComparableValue` is a fair name. With both `Less()` and `LessEqual()` on the interface, I think a user can implement any comparison they want.

Also, I hate when -`er` is appended to a non-verb or non-verb phrase. What in the world is a `Stringer` (aside from a character from HBO's The Wire) anyway?

philhassey commented on internal/eval/evalers.go:
```diff
@@ -1073,3 +1120,123 @@ func (n *ipIsInRangeEval) Eval(ctx *Context) (types.Value, error) {
+}
+
+// virtualLessEqualThanEval struct
+type virtualLessThanOrEqualEval struct {
```
I'd rather it be

`lesserLessThanOrEqualEval`

(or if we renamed to Comparable)

`comparableLessThanOrEqualEval`

apg replied:
I'll wait on this until the comparison stuff settles.

philhassey commented on internal/eval/evalers_test.go:
```diff
@@ -946,6 +946,193 @@ func TestDecimalGreaterThanOrEqualNode(t *testing.T) {
 
+// hack: newVirtualLessThanEval and friends aren't func(Evaler, Evaler) Evaler, but we can turn them into it with generics.
+func makeEvaler[T Evaler](fn func(Evaler, Evaler) T) func(Evaler, Evaler) Evaler {
+	return func(lhs Evaler, rhs Evaler) Evaler {
```
You could have them return `Evaler` then you wouldn't need this?

Commit:
Add a duration type, as per RFC 80.

This builds on the Datetime PR.

It adds the duration type, and all of the associated methods that
require durations:

- datetime.toTime
- datetime.offset
- datetime.durationSince
- duration.toMilliseconds
- duration.toSeconds
- duration.toMinutes
- duration.toHours
- duration.toDays

Signed-off-by: Andrew Gwozdziewycz <andrew.gwozdziewycz@strongdm.com>

Commit:
Address PR Feedback

- Lesser becomes ComparableValue
- Move ComparableValue to evalers only
- Move all magic values to constants
- TypeError when incompatible comparable types
- Support more deserialization for duration/datetime
- Make the datetime parser easier to follow
- Drop UnsafeDatetime in favor of FromStdTime(time.UnixMilli(..))
- Document methods
- Test Coverage to 100%

Signed-off-by: Andrew Gwozdziewycz <andrew.gwozdziewycz@strongdm.com>

Commit:
Update the README indicating datetime / duration support

Signed-off-by: Andrew Gwozdziewycz <andrew.gwozdziewycz@strongdm.com>
